graql/reasoner/query/DisjunctiveQuery.java

```diff
@@ -0,0 +1,304 @@
+                getBindingVars(),
+                traversalExecutor,
+                context()
+        );
+    }
+
+    @Override
+    ResolvableQuery copy() {
+        return new DisjunctiveQuery(
+                new HashSet<>(getClauses()),
+                getBindingVars(),
+                traversalExecutor,
+                context());
+    }
+
+    @Override
+    public boolean isAtomic() {
+        // TODO unclear of the meaning of atomicity in this case
+        return getClauses().stream().allMatch(CompositeQuery::isAtomic);
+    }
+
+    @Override
+    boolean isEquivalent(ResolvableQuery q) {
+        DisjunctiveQuery that = q.asDisjunctive();
+        return that.getClauses().size() == getClauses().size()
+                && getClauses().stream().allMatch(clause -> that.getClauses().stream().anyMatch(clause::isEquivalent));
+    }
+
+    @Override
+    public String toString(){
+        return getPattern().toString();
+    }
+
+    @Override
+    public ReasonerQuery conjunction(ReasonerQuery q) {
+        throw new UnsupportedOperationException();
+//        return new DisjunctiveQuery(
+//                getClauses().stream()
+//                        .map(CompositeQuery::getPattern)
+//                        .map(p -> Graql.and(
+//                                Sets.union(
+//                                        p.getPatterns(), getPattern().getPatterns()
+//                                )
+//                        ))
+//                        .map(p -> new CompositeQuery(
+//                                p, traversalExecutor, context()
+//                        ))
+//                        .collect(Collectors.toSet()),
+//                traversalExecutor,
+//                context()
+//        );
+    }
+
+    @Override
+    public void checkValid() {
+        clauses.forEach(CompositeQuery::checkValid);
+    }
+
+    @Override
+    public Set<Variable> getVarNames() {
+        return bindingVars;
+    }
+
+    @Override
+    public Set<Atomic> getAtoms() {
+        return getClauses().stream().flatMap(c -> c.getAtoms().stream()).collect(Collectors.toSet());
+    }
+
+    @Override
+    public Disjunction<Pattern> getPattern() {
+        // TODO How can we rebuild the disjunction in proper Graql form (with unbound variables outside the disjunction) once it's already in disjunctive normal form?!
+        return Graql.or(clauses.stream().map(CompositeQuery::getPattern).collect(Collectors.toSet()));
+    }
+
+    @Override
+    public Conjunction<Pattern> getPattern(Map<Variable, Concept> map) {
+        HashSet<Pattern> patterns = getIdPredicatePatterns(map);
+        patterns.add(getPattern());
+
+        return Graql.and(patterns);
+    }
+
+    @Override
+    public ConceptMap getSubstitution() {
+        // TODO What should this do, given that there could be a substitution for any of its clauses
```

Comment:
If there is a substitution in the query it should be in the form of id predicates. These should be converted out to match the set of `bindingVars`

Reply:
Leave as an exception for now. Will be needed if we don't normalise queries or use disjunctions in rules.